x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/AbstractCreateApiKeyRequest.java

```diff
@@ -0,0 +1,139 @@
+    }
+
+    public AbstractCreateApiKeyRequest(StreamInput in) throws IOException {
+        super(in);
+        if (in.getTransportVersion().onOrAfter(TransportVersion.V_7_10_0)) {
+            this.id = in.readString();
+        } else {
+            this.id = UUIDs.base64UUID();
+        }
+        if (in.getTransportVersion().onOrAfter(TransportVersion.V_7_5_0)) {
+            this.name = in.readOptionalString();
+        } else {
+            this.name = in.readString();
+        }
+        this.expiration = in.readOptionalTimeValue();
+        this.roleDescriptors = in.readImmutableList(RoleDescriptor::new);
+        this.refreshPolicy = WriteRequest.RefreshPolicy.readFrom(in);
+        if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_0_0)) {
+            this.metadata = in.readMap();
+        } else {
+            this.metadata = null;
+        }
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public abstract ApiKey.Type getType();
+
+    public TimeValue getExpiration() {
+        return expiration;
+    }
+
+    public List<RoleDescriptor> getRoleDescriptors() {
+        return roleDescriptors;
+    }
+
+    public WriteRequest.RefreshPolicy getRefreshPolicy() {
+        return refreshPolicy;
+    }
+
+    public Map<String, Object> getMetadata() {
+        return metadata;
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (Strings.isNullOrEmpty(name)) {
+            validationException = addValidationError("api key name is required", validationException);
+        } else {
+            if (name.length() > 256) {
+                validationException = addValidationError("api key name may not be more than 256 characters long", validationException);
+            }
+            if (name.equals(name.trim()) == false) {
+                validationException = addValidationError("api key name may not begin or end with whitespace", validationException);
+            }
+            if (name.startsWith("_")) {
+                validationException = addValidationError("api key name may not begin with an underscore", validationException);
+            }
+        }
+        if (metadata != null && MetadataUtils.containsReservedMetadata(metadata)) {
+            validationException = addValidationError(
+                "API key metadata keys may not start with [" + MetadataUtils.RESERVED_PREFIX + "]",
+                validationException
+            );
+        }
+        for (RoleDescriptor roleDescriptor : getRoleDescriptors()) {
+            validationException = RoleDescriptorRequestValidator.validate(roleDescriptor, validationException);
+        }
+        return validationException;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
```

Comment:
No because we have a new action here. Since the action won't be available in an old node, the code will fail before it even attempts to decode the request. It's like when we add any other new TransportAction, we don't need any BWC handling for its request class. 

Strictly speaking, the new CreateCrossApiKeyRequest can override this method to drop existing checks for old verions, e.g.:

```java
if (out.getTransportVersion().onOrAfter(TransportVersion.V_7_5_0)) {
    out.writeOptionalString(name);
} else {
    out.writeString(name);
}
```
But that means duplicating a bunch code (for both read and write) in subclasses. So I didn't do it since the benefit is rather marginal.

Reply:
> But that means duplicating a bunch code (for both read and write) in subclasses. So I didn't do it since the benefit is rather marginal.

On a second thought, I decided to have overridden `writeTo` and corresponding constructor for the new Request and drop these obsolete version checks. Since we are introducing a new class, might as well takes more advantage of it.